Zend/zend_ast.c

```diff
@@ -441,3 +444,1011 @@ ZEND_API void zend_ast_apply(zend_ast *ast, zend_ast_apply_func fn) {
+			}
+			break;
+		case ZEND_AST_TYPE:
+			switch (ast->attr) {
+				case IS_ARRAY:    APPEND_STR("array");
+				case IS_CALLABLE: APPEND_STR("callable");
+				default:          ZEND_ASSERT(0);
+			}
+			break;
+
+		/* 1 child node */
+		case ZEND_AST_VAR:
+			smart_str_appendc(str, '$');
+			zend_ast_export_var(str, ast->child[0], 0, indent);
+			break;
+		case ZEND_AST_CONST:
+			zv = zend_ast_get_zval(ast->child[0]);
+			smart_str_appendl(str, Z_STRVAL_P(zv), Z_STRLEN_P(zv));
+			break;
+		case ZEND_AST_RESOLVE_CLASS_NAME:
+			zend_ast_export_name(str, ast->child[0], 0, indent);
+			APPEND_STR("::class");
+		case ZEND_AST_UNPACK:
+			APPEND_NODE_1("...");
+		case ZEND_AST_UNARY_PLUS:  PREFIX_OP("+", 240, 241);
+		case ZEND_AST_UNARY_MINUS: PREFIX_OP("-", 240, 241);
+		case ZEND_AST_CAST:
+			switch (ast->attr) {
+				case IS_NULL:      PREFIX_OP("(unset)",  240, 241);
+				case _IS_BOOL:     PREFIX_OP("(bool)",   240, 241);
+				case IS_LONG:      PREFIX_OP("(int)",    240, 241);
+				case IS_DOUBLE:    PREFIX_OP("(double)", 240, 241);
+				case IS_STRING:    PREFIX_OP("(string)", 240, 241);
+				case IS_ARRAY:     PREFIX_OP("(array)",  240, 241);
+				case IS_OBJECT:    PREFIX_OP("(object)", 240, 241);
+				default:           ZEND_ASSERT(0);
+			}
+			break;
+		case ZEND_AST_EMPTY:
+			FUNC_OP("empty");
+		case ZEND_AST_ISSET:
+			FUNC_OP("isset");
+		case ZEND_AST_SILENCE:
+			PREFIX_OP("@", 240, 241);
+		case ZEND_AST_SHELL_EXEC:
+			smart_str_appendc(str, '`');
+			zend_ast_export_ex(str, ast->child[0], 0, indent);
+			smart_str_appendc(str, '`');
+			break;
+		case ZEND_AST_CLONE:
+			PREFIX_OP("clone ", 270, 271);
+		case ZEND_AST_EXIT:
+			if (ast->child[0]) {
+				FUNC_OP("exit");
+			} else {
+				APPEND_STR("exit");
+			}
+			break;
+		case ZEND_AST_PRINT:
+			PREFIX_OP("print ", 60, 61);
+		case ZEND_AST_INCLUDE_OR_EVAL:
+			switch (ast->attr) {
+				case ZEND_INCLUDE_ONCE: FUNC_OP("include_once");
+				case ZEND_INCLUDE:      FUNC_OP("include");
+				case ZEND_REQUIRE_ONCE: FUNC_OP("require_once");
+				case ZEND_REQUIRE:      FUNC_OP("require");
+				case ZEND_EVAL:         FUNC_OP("eval");
+				default:                ZEND_ASSERT(0);
+			}
+			break;
+		case ZEND_AST_UNARY_OP:
+			switch (ast->attr) {
+				case ZEND_BW_NOT:   PREFIX_OP("~", 240, 241);
+				case ZEND_BOOL_NOT: PREFIX_OP("!", 240, 241);
+				default:            ZEND_ASSERT(0);
+			}
+			break;
+		case ZEND_AST_PRE_INC:
+			PREFIX_OP("++", 240, 241);
+		case ZEND_AST_PRE_DEC:
+			PREFIX_OP("--", 240, 241);
+		case ZEND_AST_POST_INC:
+			POSTFIX_OP("++", 240, 241);
+		case ZEND_AST_POST_DEC:
+			POSTFIX_OP("--", 240, 241);
+
+		case ZEND_AST_GLOBAL:
+			APPEND_NODE_1("global ");
+		case ZEND_AST_UNSET:
+			FUNC_OP("unset");
+		case ZEND_AST_RETURN:
+			APPEND_NODE_1("return");
+		case ZEND_AST_LABEL:
+			zend_ast_export_name(str, ast->child[0], 0, indent);
+			smart_str_appendc(str, ':');
+			break;
+		case ZEND_AST_REF:
+			APPEND_NODE_1("&");
```

Comment:
Using `APPEND_NODE_1` for this causes an extra whitespace (`foreach ($a as & $b)`). Looks a bit weird, but not really a problem.